### PR TITLE
[Enhancement] opt interleaving probe to avoid bad cases

### PR DESCRIPTION
**Pull request by fzhedu**
## What I'm doing:
1. refine conditions to adaptively interleaving.
2. opt count nonzeros for int32 using SIMD.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/join_hash_map.tpp [hunk 1]
```diff
@@ -668,7 +668,10 @@ void JoinHashMap<LT, BuildFunc, ProbeFunc>::_search_ht(RuntimeState* state, Chun
     if (!_probe_state->has_remain) {
         _probe_state->probe_row_count = (*probe_chunk)->num_rows();
         _probe_state->active_coroutines = state->query_options().interleaving_group_size;
-        if (state->query_options().interleaving_group_size > 0 && !_table_items->ht_cache_miss_serious()) {
+        auto dop = state->query_options().pipeline_dop;
+        // disable interleaving for concurrent queries with lower dop if adaptively interleaving.
+        if (state->query_options().interleaving_group_size > 0 &&
+            ((dop > 0 && dop < 4) || !_table_items->ht_cache_miss_serious())) {
```
Comment (trueeyu) [inline be/src/exec/join_hash_map.tpp:674]: Can you introduce the principle of optimization?

Comment (trueeyu) [inline be/src/exec/join_hash_map.tpp:674]: Why set param should be (dop > 0 && dop < 4) ?

File: be/src/simd/simd.h [hunk 2]
```diff
@@ -53,6 +53,44 @@ inline size_t count_zero(const int8_t* data, size_t size) {
     return count;
 }
 
+inline size_t count_zero_int32(const uint32_t* data, size_t size) {
```
Comment (trueeyu) [inline be/src/simd/simd.h:56]: count_zero_uint32 ?

Comment (trueeyu) [reply be/src/simd/simd.h:56]: or cont_zero(const uint32_t* data, size_t size)?

File: be/src/simd/simd.h [hunk 3]
```diff
@@ -53,6 +53,44 @@ inline size_t count_zero(const int8_t* data, size_t size) {
...
+    // count per 16 int
+    const __m128i zero16 = _mm_setzero_si128();
+    const uint32_t* end16 = data + (size / 16 * 16);
+
+    for (; data < end16; data += 16) {
+        count += __builtin_popcountll(static_cast<uint64_t>(_mm_movemask_ps(_mm_cvtepi32_ps(_mm_cmpeq_epi32(
+                                              _mm_loadu_si128(reinterpret_cast<const __m128i*>(data)), zero16)))) |
+                                      (static_cast<uint64_t>(_mm_movemask_ps(_mm_cvtepi32_ps(_mm_cmpeq_epi32(
+                                               _mm_loadu_si128(reinterpret_cast<const __m128i*>(data + 4)), zero16))))
+                                       << 4u) |
```
Comment (ZiheLiu) [inline be/src/simd/simd.h:70]: `<< 16u`? The result of `(_mm_movemask_ps(_mm_cvtepi32_ps(_mm_cmpeq_epi32( 
                                               _mm_loadu_si128(reinterpret_cast<const __m128i*>(data + 4)), zero16))))` is 16-bit.

File: be/src/exec/join_hash_map.tpp (hunk 1 above)
Comment (fzhedu) [reply be/src/exec/join_hash_map.tpp:674]: disable interleaving for concurrent queries with lower dop (usually < 4) if adaptively interleaving.

File: be/src/simd/simd.h (hunk 3 above)
Comment (fzhedu) [reply be/src/simd/simd.h:70]: each `_mm_movemask_ps` get 4 bits results corresponding to 4 int32, so here is 4, 8, 12

File: be/src/exec/join_hash_map.tpp (hunk 1 above)
Comment (fzhedu) [reply be/src/exec/join_hash_map.tpp:674]: comments

File: be/src/simd/simd.h [hunk 4]
```diff
@@ -53,6 +53,44 @@ inline size_t count_zero(const int8_t* data, size_t size) {
...
+    return count_zero_int32(nums.data(), nums.size());
+}
+
+inline size_t count_nonzero_int32(const std::vector<uint32_t>& nums) {
+    return nums.size() - count_zero_int32(nums.data(), nums.size());
+}
+
 // Count the number of zeros of 8-bit unsigned integers.
 inline size_t count_zero(const uint8_t* data, size_t size) {
     return count_zero(reinterpret_cast<const int8_t*>(data), size);
```
[the remark this answers is not in the document]
Comment (trueeyu) [reply be/src/simd/simd.h:96]: Pay attention to the comment: Ensure data alignment when using _mm_load_si128 or use _mm_loadu_si128 for unaligned memory access

[the remark this answers is not in the document]
Comment (trueeyu) [reply be/src/simd/simd.h:96]: @fzhedu is aligned useless?

**Review by trueeyu — DISMISSED** [review]

**Review by ZiheLiu — DISMISSED** [review]

**Review by satanson — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Pull request merged**